packages/core/status/core-status-server-internal/src/log_plugins_status.ts

```
@@ -0,0 +1,185 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Observable } from 'rxjs';
+import { pairwise, takeUntil, map, filter, startWith } from 'rxjs/operators';
+import type { PluginName } from '@kbn/core-base-common';
+import { ServiceStatus, ServiceStatusLevels } from '@kbn/core-status-common';
+
+export type ServiceStatusWithName = ServiceStatus & {
+  pluginName: PluginName;
+};
+
+export type PluginStatus = 'unavailable' | 'critical' | 'degraded' | 'available';
+export type PluginsByStatus = Record<PluginStatus, ServiceStatusWithName[]>;
+
+export interface PluginStatusChanges {
+  status: PluginsByStatus;
+  updates: PluginsByStatus;
+  updated: number;
+  total: number;
+}
+
+export const getPluginsStatusChanges = (
+  plugins$: Observable<Record<PluginName, ServiceStatus>>,
+  stop$: Observable<void>
+): Observable<PluginStatusChanges> => {
+  return plugins$.pipe(
+    startWith({}), // consider all plugins unavailable by default
```

Otherwise we are silently ignoring single emission, which could be meaningful.